lib/membrane_h264_ffmpeg/parser.ex

```diff
@@ -187,37 +191,87 @@ defmodule Membrane.H264.FFmpeg.Parser do
     {:ok, %{state | parser_ref: nil}}
   end
 
-  defp parse_access_units(input, au_sizes, metadata, %{partial_frame: <<>>} = state) do
-    state = update_metadata(metadata, state)
-    {buffers, input, state} = do_parse_access_units(input, au_sizes, metadata, state, [])
+  defp parse_access_units(
+         input,
+         au_sizes,
+         metadata,
+         output_picture_numbers,
+         %{partial_frame: <<>>} = state
+       ) do
+    state = %{state | metadata: metadata}
+
+    {buffers, input, state} =
+      do_parse_access_units(input, au_sizes, metadata, output_picture_numbers, state, [])
+
     {buffers, %{state | partial_frame: input}}
   end
 
-  defp parse_access_units(input, [], _metadata, state) do
+  defp parse_access_units(input, [], _metadata, _output_picture_numbers, state) do
     {[], %{state | partial_frame: state.partial_frame <> input}}
   end
 
-  defp parse_access_units(input, [au_size | au_sizes], metadata, state) do
+  defp parse_access_units(input, [au_size | au_sizes], metadata, output_picture_numbers, state) do
     {first_au_buffers, input, state} =
-      do_parse_access_units(state.partial_frame <> input, [au_size], state.metadata, state, [])
+      do_parse_access_units(
+        state.partial_frame <> input,
+        [au_size],
+        state.metadata,
+        output_picture_numbers,
+        state,
+        []
+      )
+
+    state = %{state | metadata: metadata}
+
+    {buffers, input, state} =
+      do_parse_access_units(input, au_sizes, state.metadata, output_picture_numbers, state, [])
 
-    state = update_metadata(metadata, state)
-    {buffers, input, state} = do_parse_access_units(input, au_sizes, metadata, state, [])
     {first_au_buffers ++ buffers, %{state | partial_frame: input}}
   end
 
-  defp do_parse_access_units(input, [], _metadata, state, acc) do
+  defp do_parse_access_units(input, [], _metadata, _output_frame_numbers, state, acc) do
     {Enum.reverse(acc), input, state}
   end
 
-  defp do_parse_access_units(input, [au_size | au_sizes], metadata, state, acc) do
+  defp do_parse_access_units(
+         input,
+         [au_size | au_sizes],
+         metadata,
+         output_frame_numbers,
+         state,
+         acc
+       ) do
     <<au::binary-size(au_size), rest::binary>> = input
 
-    # setting both :timestamp and :dts in order to maintain backward compatibility
-    metadata = Map.put(metadata, :timestamp, state.timestamp) |> Map.put(:dts, state.timestamp)
+    [output_frame_number | output_frame_numbers] = output_frame_numbers
+
+    {{pts, dts}, state} =
+      if state.framerate do
+        {frames, seconds} = state.framerate
+
+        state =
+          if output_frame_number == 0,
+            do: %{
+              state
+              | frame_number_offset: state.frame_number_offset + state.last_frame_number
```

Comment:
Can't we keep this offset on the C side and return already adjusted frame numbers?

Reply:
Nice point, I'll change that.